src/Data/Parameterized/FinMap/Safe.hs

```diff
@@ -0,0 +1,209 @@
+{-|
+Copyright        : (c) Galois, Inc 2022
+
+See "Data.Parameterized.FinMap".
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Parameterized.FinMap.Safe
+  ( FinMap
+  -- * Query
+  , null
+  , lookup
+  , size
+  -- * Construction
+  , incMax
+  , embed
+  , empty
+  , singleton
+  , insert
+  , buildFinMap
+  , append
+  , fromVector
+  -- * Operations
+  , delete
+  , decMax
+  , mapWithKey
+  ) where
+
+import           Prelude hiding (lookup, null)
+
+import           Data.Foldable.WithIndex (FoldableWithIndex(ifoldMap))
+import           Data.Functor.WithIndex (FunctorWithIndex(imap))
+import           Data.Maybe (isJust)
+import           Data.Proxy (Proxy(Proxy))
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           GHC.TypeLits (KnownNat, Nat)
+
+import           Data.Parameterized.Fin (Fin)
+import qualified Data.Parameterized.Fin as Fin
+import           Data.Parameterized.NatRepr (NatRepr, type (+), type (<=))
+import qualified Data.Parameterized.NatRepr as NatRepr
+import           Data.Parameterized.Vector (Vector)
+import qualified Data.Parameterized.Vector as Vec
+
+------------------------------------------------------------------------
+-- Type
+
+-- | @'FinMap' n a@ is a map with @'Fin' n@ keys and @a@ values.
+data FinMap (n :: Nat) a =
+  FinMap
+    { getFinMap :: Map (Fin n) a
+    , maxSize :: NatRepr n
+    }
+
+instance Eq a => Eq (FinMap n a) where
```

Comment:
> I suspect this is a common requirement to using this instance, so maybe it would be helpful to add a finMapEq function here that encapsulates that.

To be sure I understand - what would be the type of `finMapEq`, and what would it do? Is it one of the following?
```haskell
finMapEq :: FinMap n a -> FinMap n a -> Bool
finMapEq :: FinMap n a -> FinMap m a -> Bool
```

Reply:
Ah, so then we get into some interesting questions - The unsafe interface can't *check* whether `n ~ m`, because it doesn't have a `NatRepr` around. However, both interfaces would be able to check whether the maps had the same (key, value) pairs - modulo that the safe implementation would cast its `Fin n`s to `Natural`s to perform this check, since you also can't compare `Fin n` with `Fin m` directly. I'm open to implementing that - let's see if it comes up in UC-Crux.